src/js/tracks/text-track-list.js

```diff
@@ -18,131 +19,145 @@ import document from 'global/document';
  *   attribute EventHandler onaddtrack;
  *   attribute EventHandler onremovetrack;
  * };
+ *
+ * @param {Track[]} tracks A list of tracks to add to the list
+ * @extends EventTarget
+ * @class TextTrackList
  */
-function TextTrackList (tracks) {
-  let list = this;
 
-  if (browser.IS_IE8) {
-    list = document.createElement('custom');
+class TextTrackList extends EventTarget {
+  constructor(tracks = []) {
+    super();
+    let list = this;
 
-    for (let prop in TextTrackList.prototype) {
-      if (prop !== 'constructor') {
-        list[prop] = TextTrackList.prototype[prop];
+    if (browser.IS_IE8) {
+      list = document.createElement('custom');
+
+      for (let prop in TextTrackList.prototype) {
+        if (prop !== 'constructor') {
+          list[prop] = TextTrackList.prototype[prop];
+        }
       }
     }
-  }
 
-  tracks = tracks || [];
-  list.tracks_ = [];
+    list.tracks_ = [];
 
-  Object.defineProperty(list, 'length', {
-    get: function() {
-      return this.tracks_.length;
-    }
-  });
+    Object.defineProperty(list, 'length', {
+      get() {
+        return this.tracks_.length;
+      }
+    });
 
-  for (let i = 0; i < tracks.length; i++) {
-    list.addTrack_(tracks[i]);
-  }
+    for (let i = 0; i < tracks.length; i++) {
+      list.addTrack_(tracks[i]);
+    }
 
-  if (browser.IS_IE8) {
-    return list;
+    if (browser.IS_IE8) {
+      return list;
+    }
   }
-}
 
-TextTrackList.prototype = Object.create(EventTarget.prototype);
-TextTrackList.prototype.constructor = TextTrackList;
-
-/*
- * change - One or more tracks in the track list have been enabled or disabled.
- * addtrack - A track has been added to the track list.
- * removetrack - A track has been removed from the track list.
- */
-TextTrackList.prototype.allowedEvents_ = {
-  'change': 'change',
-  'addtrack': 'addtrack',
-  'removetrack': 'removetrack'
-};
+  /**
+   * Add TextTrack from TextTrackList
+   *
+   * @param {TextTrack} track
+   * @method addTrack_
+   * @private
+   */
+  addTrack_(track) {
+    let index = this.tracks_.length;
+
+    if (!('' + index in this)) {
+      Object.defineProperty(this, index, {
+        get() {
+          return this.tracks_[index];
+        }
+      });
+    }
 
-// emulate attribute EventHandler support to allow for feature detection
-for (let event in TextTrackList.prototype.allowedEvents_) {
-  TextTrackList.prototype['on' + event] = null;
-}
+    track.addEventListener('modechange', Fn.bind(this, function() {
+      this.trigger('change');
+    }));
+    this.tracks_.push(track);
 
-/**
- * Add TextTrack from TextTrackList
- *
- * @param {TextTrack} track
- * @method addTrack_
- * @private
- */
-TextTrackList.prototype.addTrack_ = function(track) {
-  let index = this.tracks_.length;
-  if (!(''+index in this)) {
-    Object.defineProperty(this, index, {
-      get: function() {
-        return this.tracks_[index];
-      }
+    this.trigger({
+      type: 'addtrack',
+      track
```

Comment:
Would storing the array inside of the TextTrackList object rather than tracks_ work for this? or do we not want to remove that variable? This is the idea

``` JavaScript
add(track) {
    // auto increments length for us and stores it in our object under the correct index
    [].push.call(this, track);
  }
  remove(track) {
    for(let i = 0; i < this.length; i++) {
      if(this[i] === track) {
        // auto decrements length for us and splices right out of our object
        [].splice.call(this, i, 1);
        break;
      }
    }
  }

```